README.md
# YouTube Downloader
Este é um aplicativo simples feito em Python com a biblioteca 
PySimpleGUI e o pytube que permite baixar vídeos MP4 ou áudios MP3 de um 
link informado pelo usuário.
Requisitos
Certifique-se de ter o Python instalado em seu sistema.

## Instalação das dependências
Para instalar as dependências necessárias, siga as etapas abaixo:
1. Faça o download ou clone este repositório em seu sistema.
2. Abra o terminal ou prompt de comando e navegue até o diretório onde 
você salvou o aplicativo.
3. Execute o seguinte comando para instalar as dependências:
``` bash
# Esse comando irá ler o arquivo requirements.txt e instalar todas as bibliotecas necessárias.
pip install -r requirements.txt
```

## Como executar o aplicativo
1. Certifique-se de ter instalado todas as dependências mencionadas 
acima.
2. Abra o terminal ou prompt de comando e navegue até o diretório onde 
você salvou o aplicativo.
3. Execute o seguinte comando para iniciar o aplicativo:
4. python app.py
5. A interface gráfica será exibida. Informe o link do vídeo ou áudio 
que você deseja baixar e selecione o tipo de mídia (Vídeo MP4 ou Áudio 
MP3).
6. Clique no botão "Baixar" para iniciar o processo de download.
7. Após o download ser concluído, uma mensagem de sucesso será exibida.

## Observações
 O arquivo baixado será salvo com um nome padrão (downloaded_media.mp4 
 ou downloaded_media.mp3) dentro da pasta downloads e na pasta adequada 
 para audio ou video, dentro do mesmo diretório onde o script está 
 localizado. Caso você queira alterar o nome ou local do arquivo, você 
 pode editar a variável output_path no código.
 Certifique-se de ter permissões de escrita no diretório onde o script 
 está localizado, pois é onde o arquivo baixado será salvo.
 Este aplicativo utiliza o pytube como biblioteca para fazer o download 
 dos vídeos/áudios. Certifique-se de que o link fornecido seja 
 compatível com o pytube.

Espero que este aplicativo seja útil para você! Se você tiver alguma 
dúvida ou sugestão de melhoria, sinta-se à vontade para entrar em 
contato através do e-mail 
[suporte.ryu@gmail.com](mailto:suporte.ryu@gmail.com)!


CHANGELOG
# YouTube Downloader Release Notes

## 0.0.1 (17 Jun 2023)

* Initial release


src/youtube_downloader/app.py
"""
An Youtube downloader
"""
import toga
from toga.style import Pack
from toga.style.pack import COLUMN, ROW

class YouTubeDownloader(toga.App):

    def startup(self):
        """
        Construct and show the Toga application.

        Usually, you would add your application to a main content box.
        We then create a main window (with a name matching the app), and
        show the main window.
        """
        
        search_box = self.construct_search_box()
        preview_box = self.construct_preview_box()
        download_box = self.construct_download_box()

        main_box = toga.Box(
            children=[
                search_box,
                preview_box,
                toga.Divider(),
                download_box
            ],
            style=Pack(direction=COLUMN)
        )

        self.main_window = toga.MainWindow(title=self.formal_name)
        self.main_window.content = main_box
        self.main_window.show()
        self.thumb_view.refresh()

    def construct_search_box(self):
        search_button = toga.Button(
            "Search",
            on_press=self.get_stream,
            style=Pack(padding=5)
        )
        self.search_input = toga.TextInput(style=Pack(flex=1))

        search_box = toga.Box(style=Pack(direction=ROW, padding=5))
        search_box.add(search_button)
        search_box.add(self.search_input)

        return search_box
    
    def construct_preview_box(self):
        thumb_scale = .5
        thumb_size = {"width":1280*thumb_scale, "height":720*thumb_scale}
        thumb_image = toga.Image(path="resources/images/thumbnail_placeholder.jpg") # /1280x720
        self.thumb_view = toga.ImageView(
            id='thumb',
            image=thumb_image, 
            style=Pack(width=thumb_size["width"], height=thumb_size["height"])
        )

        stream_title_label = toga.Label("Title: ")
        self.stream_title = toga.Label("")

        title_box = toga.Box(style=Pack(direction=ROW, padding=5))
        title_box.add(stream_title_label)
        title_box.add(self.stream_title)

        stream_author_label = toga.Label("Author: ")
        self.stream_author = toga.Label("")

        author_box = toga.Box(style=Pack(direction=ROW, padding=5))
        author_box.add(stream_author_label)
        author_box.add(self.stream_author)

        stream_length_label = toga.Label("Length: ")
        self.stream_length = toga.Label("")

        length_box = toga.Box(style=Pack(direction=ROW, padding=5))
        length_box.add(stream_length_label)
        length_box.add(self.stream_length)

        preview_box = toga.Box(
            children=[
                self.thumb_view,
                title_box,
                author_box,
                length_box,
            ],
            style=Pack(direction=COLUMN, padding=5)
        )

        return preview_box

    def construct_download_box(self):
        download_mp4_button = toga.Button(
            "Download mp4",
            on_press=self.download_as_mp4,
            style=Pack(padding=5)
        )
        
        download_mp3_button = toga.Button(
            "Download mp3",
            on_press=self.download_as_mp3,
            style=Pack(padding=5)
        )

        download_box = toga.Box(style=Pack(direction=ROW, padding=5))
        download_box.add(download_mp4_button)
        download_box.add(download_mp3_button)

        return download_box

    def get_stream(self, widget):
        print(f"Search link: '{self.search_input.value}'")
        self.stream_title.text = f"Title of link: '{self.search_input.value}'"
        self.stream_author.text = f"Author of link: '{self.search_input.value}'"
        self.stream_length.text = f"Length of link: '{self.search_input.value}'"

    def download_as_mp4(self, widget):
        print(f"Download link '{self.search_input.value}' as mp4")

    def download_as_mp3(self, widget):
        print(f"Download link '{self.search_input.value}' as mp3")

def main():
    return YouTubeDownloader()


main.py
import os
import requests
from io import BytesIO
from PIL import Image
from urllib.parse import urlparse, urlencode, parse_qs, urlunparse


import PySimpleGUI as sg
from pytube import YouTube

def remove_url_parameter(url, parameter):
    parsed_url = urlparse(url)
    query_params = parse_qs(parsed_url.query)

    if parameter in query_params:
        del query_params[parameter]

    updated_query_string = urlencode(query_params, doseq=True)
    parsed_url = parsed_url._replace(query=updated_query_string)

    new_url = urlunparse(parsed_url)
    return new_url

def normalize_youtube_url(url):
    url = remove_url_parameter(url, 't')
    if not url.startswith('https'):
        if not url.startswith('www.'):
           url = f'www.{url}'
        url = f'https://{url}'
    if 'youtu.be' in url:
        video_id = url.split('/')[-1]
        new_url = f'https://www.youtube.com/watch?v={video_id}'
        return new_url
    return url    

def get_image_from_url(url):
    response = requests.get(url)
    image = Image.open(BytesIO(response.content))
    ratio = 1 / 2
    image.thumbnail((int(720 * ratio), int(1280 * ratio)))
    thumb_data = BytesIO()
    image.save(thumb_data, format="PNG")
    thumb_data.seek(0)
    return thumb_data.read()

def progress_func(stream, chunk, bytes_remaining):
    print('Realizando download...')
    print(f'Stream filesize: {stream.filesize_mb} MB')
    print(f'Bytes remaining: {bytes_remaining} bytes')
    # print(f'Length chunk: {chunk}')
    print('----')
    
def complete_func(stream, *args):
    print(f'Download {stream.title} completo')

def get_video(link):
    youtube_video = YouTube(
    	link,
    	on_progress_callback=progress_func,
        on_complete_callback=complete_func,
        use_oauth=False,
        allow_oauth_cache=True
    )
    return youtube_video

def download_audio(link):
    download_audio_dir = './downloads/audio/'
    youtube_video = get_video(link)
    stream = youtube_video.streams.filter(only_audio=True).first()
    print(f'Baixando {youtube_video.title}')
    try:
        out_file = stream.download(output_path=download_audio_dir)
        base, ext = os.path.splitext(out_file)
        filename = base + '.mp3'
        os.rename(out_file, filename)
        print("Download concluído com sucesso!")
    except Exception as e:
        print(f"Ocorreu um erro durante o download: {e}")
    
def download_video(link):
    download_video_dir = './downloads/video/'
    youtube_video = get_video(link)
    stream = youtube_video.streams.filter(file_extension='mp4').first()
    print(f'Baixando {youtube_video.title}')
    try:
        out_file = stream.download(output_path=download_video_dir)
        print("Download concluído com sucesso!")
    except Exception as e:
        print(f"Ocorreu um erro durante o download: {e}")

def format_duration(seconds):
    hours = seconds // 3600
    minutes = (seconds % 3600) // 60
    seconds = seconds % 60
    if hours > 0:
        return f"{hours:02d}:{minutes:02d}:{seconds:02d}"
    else:
        return f"{minutes:02d}:{seconds:02d}"

sg.theme('Purple')
link = 'https://www.youtube.com/watch?v=eWWWHilDBx8'

download_layout = [
    [sg.Text('Selecione o tipo de mídia:')],
    [sg.Radio('Vídeo MP4', 'MEDIA_TYPE', key='-VIDEO-'), sg.Radio('Áudio MP3', 'MEDIA_TYPE', default=True, key='-AUDIO-')],
    [sg.Button('Baixar'), sg.Button('Sair')]
]

search_layout = [
    [sg.Input(link, key='-URL-'), sg.Button('Buscar')],
    [sg.Text('Title: '),sg.Text(key='-TITLE-')],
    [sg.Text('Author: '),sg.Text(key='-AUTHOR-')],
    [sg.Text('Length: '),sg.Text(key='-LENGTH-')],
    [sg.Combo(['mp3', 'mp4'], key='-FORMAT-')],
    [sg.Image(key='-IMAGE-')],
]

layout = [[
    sg.Column(search_layout),
    sg.VSeperator(),
    sg.Column(download_layout),
]]

window = sg.Window('Download de Vídeo/Áudio', layout)

while True:
    event, values = window.read()
    if event in (sg.WINDOW_CLOSED, 'Sair'):
        break
    if event == 'Buscar':
        url = values['-URL-']
        print(url)
        yt = YouTube(url)
        window['-TITLE-'].update(yt.title)
        window['-AUTHOR-'].update(yt.author)
        window['-LENGTH-'].update(format_duration(yt.length))
        thumb_data = get_image_from_url(yt.thumbnail_url)
        window['-IMAGE-'].update(data=thumb_data)
    if event == 'Baixar':
        url = values['-URL-']
        media_type = 'video' if values['-VIDEO-'] else 'audio'
        if (media_type == 'video'):
            download_video(url)
        else:
            download_audio(url)
        sg.popup(f'{media_type.capitalize()} baixado com sucesso!')

window.close()

